### PR TITLE
escape: Ensure that output is always valid UTF-8

### PR DESCRIPTION
**Pull request by msizanoen1**
This ensures that shell string escape operations will not produce output with invalid UTF-8 from the input by escaping invalid UTF-8 data as if they were single byte characters.

### CODE REVIEW

File: src/basic/escape.c [hunk 1]
```diff
@@ -474,14 +474,18 @@ char* octescape(const char *s, size_t len) {
...
+                int l = utf8_encoded_valid_unichar(s, SIZE_MAX);
+
+                if (char_is_cc(*s) || l < 0)
                         t += cescape_char(*s, t);
                 else {
                         if (*s == '\\' || strchr(bad, *s))
                                 *(t++) = '\\';
-                        *(t++) = *s;
+                        memcpy(t, s, l);
+                        t += l;
```
Comment (poettering) [inline src/basic/escape.c]: hmm, wouldn't it make sense to rewrite this like this:

```c
…
else if (l == 1) {
        if (*s == '\\' || strchr(bad, *s))
                *(t++) = '\\';
        *(t++) = *s;
} else {
        memcpy(t, s, l);
        t += l;
}
…
```

i.e. do the "bad char" handling only if l == 1, i.e. for regular ascii chars? that sounds a bit more correct to me

Comment (msizanoen1) [reply src/basic/escape.c]: Done.

File: src/basic/escape.c [hunk 2]
```diff
@@ -474,14 +474,21 @@ char* octescape(const char *s, size_t len) {
 static char* strcpy_backslash_escaped(char *t, const char *s, const char *bad) {
         assert(bad);
 
-        for (; *s; s++)
-                if (char_is_cc(*s))
-                        t += cescape_char(*s, t);
-                else {
+        while (*s) {
```
Comment (bluca) [inline src/basic/escape.c:479]: `while (s && *s)`

File: src/basic/escape.c [hunk 3]
```diff
@@ -474,14 +474,21 @@ char* octescape(const char *s, size_t len) {
 static char* strcpy_backslash_escaped(char *t, const char *s, const char *bad) {
         assert(bad);
```
Comment (bluca) [inline src/basic/escape.c:478]: `assert(t)`

File: src/basic/escape.c (hunk 2 above)
Comment (msizanoen1) [reply src/basic/escape.c:479]: The current code already assumes that `t` and `s` are non-`NULL`. Any changes to this, if necessary, should be done in another PR.

**Comment by medhefgo** [comment]
I assume you have a specific issue you came across? Could you add a testcase for this?

File: src/basic/escape.c (hunk 2 above)
Comment (poettering) [reply src/basic/escape.c:479]: please add assert(t) and assert(s) to the top. we generally have this rule that when you touch code and it isn't really up to our current coding style, then you fix it while touching it. Here it means encoding expectations explicitly via assert() or similar. Hence please add the two lines, either into this commit or as separate commits in your PR.

assert()s have three uses: they are documentation about expectations, they are safety nets to catch unexpected programming errors, and they are nowadays used by compilers and static analyzers to validate and optimize code, as it encodes in computer readable ways what state is possible and what is deemed impossible.

File: src/basic/escape.c [hunk 4]
```diff
@@ -474,14 +474,21 @@ char* octescape(const char *s, size_t len) {
...
+                if (char_is_cc(*s) || l < 0)
+                        t += cescape_char(*(s++), t);
+                else if (l == 1) {
                         if (*s == '\\' || strchr(bad, *s))
                                 *(t++) = '\\';
-                        *(t++) = *s;
+                        *(t++) = *(s++);
+                } else {
+                        memcpy(t, s, l);
+                        t += l;
```
Comment (poettering) [inline src/basic/escape.c]: you could replace these two lines by:

```
t = mempcpy(t, s, l);
```

Comment (msizanoen1) [reply src/basic/escape.c]: Done.

File: src/basic/escape.c (hunk 3 above)
Comment (msizanoen1) [reply src/basic/escape.c:478]: Done.

File: src/basic/escape.c (hunk 2 above)
Comment (msizanoen1) [reply src/basic/escape.c:479]: Done.

**Pull request merged**